File: apps/frontend/lib/api/tailor.ts
/**
 * Tailor pipeline API client.
 *
 * Three endpoints:
 *   POST /resumes/{id}/tailor        — start pipeline
 *   GET  /resumes/{id}/tailor/status — poll for progress
 *   POST /resumes/{id}/tailor/cancel — cancel
 */

import { apiFetch, readApiErrorMessage } from './client';

export type TailorStatus = 'running' | 'completed' | 'failed' | 'canceled';

export type PromptProfileId = 'profile1' | 'profile2' | 'profile3' | 'profile4';

export interface TailorStartResponse {
  job_id: string;
  status: string;
  message: string;
}

export interface TailorStatusResponse {
  job_id: string;
  status: TailorStatus;
  progress_stage: string | null;
  prompt_profile_id: PromptProfileId;
  started_at: string;
  completed_at: string | null;
  tailored_resume_id: string | null;
  error_message: string | null;
}

export interface StartTailorRequest {
  prompt_profile_id: PromptProfileId;
  jd_url?: string | null;
  jd_text?: string | null;
}

export async function startTailor(
  resumeId: string,
  request: StartTailorRequest
): Promise<TailorStartResponse> {
  const resp = await apiFetch(`/api/v1/resumes/${resumeId}/tailor`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(request),
  });
  if (!resp.ok) {
    const msg = await readApiErrorMessage(resp, 'Failed to start tailor job.');
    throw new Error(msg);
  }
  return resp.json() as Promise<TailorStartResponse>;
}

export async function getTailorStatus(resumeId: string): Promise<TailorStatusResponse> {
  const resp = await apiFetch(`/api/v1/resumes/${resumeId}/tailor/status`, {
    method: 'GET',
  });
  if (!resp.ok) {
    const msg = await readApiErrorMessage(resp, 'Failed to get tailor status.');
    throw new Error(msg);
  }
  return resp.json() as Promise<TailorStatusResponse>;
}

export async function cancelTailor(resumeId: string): Promise<void> {
  const resp = await apiFetch(`/api/v1/resumes/${resumeId}/tailor/cancel`, {
    method: 'POST',
  });
  if (!resp.ok) {
    const msg = await readApiErrorMessage(resp, 'Failed to cancel tailor job.');
    throw new Error(msg);
  }
}

/** Human-readable label for each pipeline progress stage. */
export function progressStageLabel(stage: string | null): string {
  switch (stage) {
    case 'apify':
      return 'Extracting job description…';
    case 'prompt1':
      return 'Analyzing job requirements…';
    case 'prompt2':
      return 'Building resume strategy…';
    case 'prompt3':
      return 'Writing tailored resume…';
    case 'postprocess':
      return 'Finalizing resume…';
    case 'starting':
      return 'Starting…';
    default:
      return 'Working…';
  }
}

File: apps/backend/app/schemas/tailor.py
"""Pydantic schemas for tailor pipeline endpoints."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, field_validator

VALID_PROMPT_PROFILES = {
    "profile1",
    "profile2",
    "profile3",
    "profile4",
    "profile5",
}


class TailorRequest(BaseModel):
    prompt_profile_id: str = "profile5"
    jd_url: str | None = None
    jd_text: str | None = None

    @field_validator("jd_url", "jd_text", mode="before")
    @classmethod
    def empty_str_to_none(cls, v: Any) -> Any:
        if isinstance(v, str) and not v.strip():
            return None
        return v

    @field_validator("prompt_profile_id")
    @classmethod
    def validate_profile(cls, v: str) -> str:
        if v not in VALID_PROMPT_PROFILES:
            raise ValueError(
                f"Invalid prompt profile: {v!r}. Must be one of {sorted(VALID_PROMPT_PROFILES)}"
            )
        return v


class TailorStartResponse(BaseModel):
    job_id: str
    status: str
    message: str


class TailorStatusResponse(BaseModel):
    job_id: str
    status: str  # "running" | "completed" | "failed" | "canceled"
    progress_stage: str | None = None
    prompt_profile_id: str
    started_at: str
    completed_at: str | None = None
    tailored_resume_id: str | None = None
    error_message: str | None = None
